sdk/nodejs/runtime/invoke.ts

```diff
@@ -100,3 +128,135 @@ export async function invoke(tok: string, props: Inputs, opts?: InvokeOptions):
+
+    // Encode the request.
+    const reqBytes = Buffer.from(req.serializeBinary());
+
+    // Write the request length.
+    const reqLen = Buffer.alloc(4);
+    reqLen.writeUInt32BE(reqBytes.length, 0);
+    fs.writeSync(syncInvokes.requests, reqLen);
+    fs.writeSync(syncInvokes.requests, reqBytes);
+
+    // Read the response.
+    const respLenBytes = Buffer.alloc(4);
+    fs.readSync(syncInvokes.responses, respLenBytes, 0, 4, null);
+    const respLen = respLenBytes.readUInt32BE(0);
+    const respBytes = Buffer.alloc(respLen);
+    fs.readSync(syncInvokes.responses, respBytes, 0, respLen, null);
+
+    // Decode the response.
+    const resp = providerproto.InvokeResponse.deserializeBinary(new Uint8Array(respBytes));
+
+    // If there were failures, propagate them.
+    processPotentialFailures(tok, resp);
+
+    // Finally propagate any other properties that were given to us as outputs.
+    const resultValue = deserializeProperties(resp.getReturn());
+
+    // Expose the properties of the actual result of invoke directly on the promise itself.
+    const promise = Promise.resolve(resultValue);
+    Object.assign(promise, resultValue);
+
+    return promise;
+}
+
+function getProvider(tok: string, opts: InvokeOptions) {
+    return opts.provider ? opts.provider :
+           opts.parent ? opts.parent.getProvider(tok) : undefined;
+}
+
+function serializePropertiesSync(prop: any): any {
+    if (prop === undefined ||
+        prop === null ||
+        typeof prop === "boolean" ||
+        typeof prop === "number" ||
+        typeof prop === "string") {
+
+        return prop;
+    }
+
+    if (asset.Asset.isInstance(prop) || asset.Archive.isInstance(prop)) {
+        throw new Error("Assets and Archives cannot be passed in as arguments to a data source call.");
+    }
+
+    if (prop instanceof Promise) {
+        throw new Error("Promises cannot be passed in as arguments to a data source call.");
+    }
+
+    if (Output.isInstance(prop)) {
+        throw new Error("Outputs cannot be passed in as arguments to a data source call.");
+    }
+
+    if (Resource.isInstance(prop)) {
+        throw new Error("Resources cannot be passed in as arguments to a data source call.");
+    }
+
+    if (prop instanceof Array) {
+        const result: any[] = [];
+        for (let i = 0; i < prop.length; i++) {
+            // When serializing arrays, we serialize any undefined values as `null`. This matches JSON semantics.
+            const elem = serializePropertiesSync(prop[i]);
+            result.push(elem === undefined ? null : elem);
+        }
+        return result;
+    }
+
+    return serializeAllKeys(prop, {});
+
+    function serializeAllKeys(innerProp: any, obj: any) {
+        for (const k of Object.keys(innerProp)) {
+            // When serializing an object, we omit any keys with undefined values. This matches JSON semantics.
+            const v = serializePropertiesSync(innerProp[k]);
+            if (v !== undefined) {
```

Comment:
preserving behavior/comments here as well between serializeProperties and serializePropertiesSync